base/special/trig.jl

```diff
@@ -22,65 +22,328 @@ end
 ## software is freely granted, provided that this notice
 ## is preserved.
 
-function sin_kernel(x::DoubleFloat64)
-    S1 = -1.66666666666666324348e-01
-    S2 =  8.33333333332248946124e-03
-    S3 = -1.98412698298579493134e-04
-    S4 =  2.75573137070700676789e-06
-    S5 = -2.50507602534068634195e-08
-    S6 =  1.58969099521155010221e-10
-
-    z = x.hi*x.hi
-    w = z*z
-    r = S2+z*(S3+z*S4) + z*w*(S5+z*S6)
-    v = z*x.hi
-    x.hi-((z*(0.5*x.lo-v*r)-x.lo)-v*S1)
+# Trigonometric functions
+# sin methods
+@noinline sin_domain_error(x) = throw(DomainError(x, "sin(x) is only defined for finite x."))
+@inline function sin(x::T) where T<:Union{Float32, Float64}
+    absx = abs(x)
+    if absx < T(pi)/4 #|x| ~<= pi/4, no need for reduction
+        if absx < sqrt(eps(T))
+            return x
+        end
+        return sin_kernel(x)
+    elseif isnan(x)
+        return T(NaN)
+    elseif isinf(x)
+        sin_domain_error(x)
+    end
+    n, y = rem_pio2_kernel(x)
+    n = n&3
+    if n == 0
+        return sin_kernel(y)
+    elseif n == 1
+        return cos_kernel(y)
+    elseif n == 2
+        return -sin_kernel(y)
+    else
+        return -cos_kernel(y)
+    end
 end
 
-function cos_kernel(x::DoubleFloat64)
-    C1 =  4.16666666666666019037e-02
-    C2 = -1.38888888888741095749e-03
-    C3 =  2.48015872894767294178e-05
-    C4 = -2.75573143513906633035e-07
-    C5 =  2.08757232129817482790e-09
-    C6 = -1.13596475577881948265e-11
+# Coefficients in 13th order polynomial approximation on [0; π/4]
+#     sin(x) ≈ x + S1*x³ + S2*x⁵ + S3*x⁷ + S4*x⁹ + S5*x¹¹ + S6*x¹³
+# D for double, S for sin, number is the order of x-1
 
-    z = x.hi*x.hi
-    w = z*z
-    r = z*(C1+z*(C2+z*C3)) + w*w*(C4+z*(C5+z*C6))
-    hz = 0.5*z
-    w = 1.0-hz
-    w + (((1.0-w)-hz) + (z*r-x.hi*x.lo))
+const DS1 = -1.66666666666666324348e-01
+const DS2 = 8.33333333332248946124e-03
+const DS3 = -1.98412698298579493134e-04
+const DS4 = 2.75573137070700676789e-06
+const DS5 = -2.50507602534068634195e-08
+const DS6 = 1.58969099521155010221e-10
+"""
+    sin_kernel(yhi, ylo)
+
+Computes the sine on the interval [-π/4; π/4].
+"""
+@inline function sin_kernel(y::DoubleFloat64)
+    y² = y.hi*y.hi
+    y⁴ = y²*y²
+    r  = @horner(y², DS2, DS3, DS4) + y²*y⁴*@horner(y², DS5, DS6)
+    y³ = y²*y.hi
+    y.hi-((y²*(0.5*y.lo-y³*r)-y.lo)-y³*DS1)
+end
+@inline function sin_kernel(y::Float64)
+    y² =  y*y
+    y⁴ =  y²*y²
```

Comment:
It is a little bit confusing to me why y^2 *y^2 is called y.

Reply:
there's  a small 4 super script there... maybe I should ditch the unicode?

Reply:
Oh, it was completely missing on my phone.

Reply:
I will say, that even on my computer, it's very small here on github :) It's just easier to follow the code than names like `x` `y` `z` `w` for all kinds of powers of `y` imo, but it's of course a problem if it's hard to read other places than my own editor :)

Reply:
I'm fine with keeping it.